data/data-pipeline/data_pipeline/comparison_tool/src/utils.py

```diff
@@ -0,0 +1,357 @@
+    return (
+        full_df[full_df[grouping_column]][demo_columns]
+        .mean()
+        .T.rename(grouping_column)
+    )
+
+
+def get_tract_level_grouping(
+    joined_df: pd.DataFrame,
+    score_column: str,
+    comparator_column: str,
+    demo_columns: list,
+    keep_missing_values: bool = True,
+) -> pd.DataFrame:
+    """Function to produce segmented statistics (tract level)
+
+    Here, we are thinking about the following segments:
+    1. CEJST and comparator
+    2. Not CEJST and comparator
+    3. Not CEJST and not comparator
+    4. CEJST and not comparator
+
+    If "keep_missing_values" flag:
+    5. Missing from CEJST and comparator (this should never be true - it would be a tract we had not seen in CEJST!)
+    6. Missing from comparator and not highlighted by CEJST
+    7. Missing from comparator and highlighted by CEJST
+
+    This will make sure that comparisons are "apples to apples".
+    """
+    group_list = [score_column, comparator_column]
+    use_df = joined_df.copy()
+    if keep_missing_values:
+        use_df = use_df.fillna({score_column: "nan", comparator_column: "nan"})
+    grouping_df = use_df.groupby(group_list)[demo_columns].mean().reset_index()
+    # this will work whether or not there are "nans" present
+    grouping_df[score_column] = grouping_df[score_column].map(
+        {
+            True: "CEJST",
+            False: "Not CEJST",
+            "nan": "No CEJST classification",
+        }
+    )
+    grouping_df[comparator_column] = grouping_df[comparator_column].map(
+        {
+            True: "Comparator",
+            False: "Not Comparator",
+            "nan": "No Comparator classification",
+        }
+    )
+    return grouping_df.set_index([score_column, comparator_column]).T
+
+
+def format_multi_index_for_excel(
+    df: pd.DataFrame, rename_str: str = "Variable"
+) -> pd.DataFrame:
+    """Helper function for multiindex printing"""
+    df = df.reset_index()
+    df.columns = [rename_str] + [
+        ", ".join(col_tuple).strip()
+        for col_tuple in df.columns[1:].to_flat_index()
+    ]
+    return df
+
+
+def get_final_summary_info(
+    population: pd.DataFrame,
+    comparator_file: str,
+    geoid_col: str,
+) -> tuple[pd.DataFrame, str]:
+    """
+    This creates a series that tells us what share (%) of census tracts identified
+    by the comparator are also in CEJST and what states the comparator covers.
+    """
+    try:
+        comparator_and_cejst_proportion_series = (
+            population.loc[(True, True)] / population.loc[(True,)].sum()
+        )
+    except KeyError:
+        # for when we are looking at a disjoint set, like donut holes
+        comparator_and_cejst_proportion_series = pd.DataFrame()
+
+    states_represented = (
+        pd.read_csv(
+            comparator_file, usecols=[geoid_col], dtype={geoid_col: str}
+        )[geoid_col]
+        .str[:2]
+        .unique()
+    )
+    states = ", ".join(
```

Comment:
Could you add a comment explaining this? This is the type of thing where I can spend 1-2 minutes figuring out why you did this, but easier if you just write a comment.

Reply:
I'm also still not sure why we're doing this. Do the territories not exist in FIPS_MAP? Why not? Can we just add them there or do you have some other objective?

Reply:
i didn't want to drop any erroneous fips codes from a comparator artificially.

Reply:
I've updated this so it's hopefully clearer!